Data/SBV/Core/Symbolic.hs

```diff
@@ -880,7 +881,7 @@ type UIMap     = Map.Map String SBVType
 type CgMap     = Map.Map String [String]
 
 -- | Cached values, implementing sharing
-type Cache a   = IMap.IntMap [(StableName (State -> IO a), a)]
+type Cache a   = HMap.HashMap (StableName (State -> IO a))  a
```

Comment:
This doesn't seem to handle the (rare but possible case) of two values hashing to the same value. The list is intentional here. Perhaps you can replace it with `HMap.HashMap (StableName (State -> IO a)) [a]`?